src/main/java/org/opensearch/neuralsearch/processor/InferenceProcessor.java

```diff
@@ -317,21 +325,45 @@ private void buildMapWithProcessorKeyAndOriginalValueForMapType(
                     List<Object> listOfStrings = list.stream().map(x -> x.get(nestedFieldMapEntry.getKey())).collect(Collectors.toList());
                     Map<String, Object> map = new LinkedHashMap<>();
                     map.put(nestedFieldMapEntry.getKey(), listOfStrings);
-                    buildMapWithProcessorKeyAndOriginalValueForMapType(
-                        nestedFieldMapEntry.getKey(),
-                        nestedFieldMapEntry.getValue(),
-                        map,
-                        next
-                    );
+                    buildNestedMap(nestedFieldMapEntry.getKey(), nestedFieldMapEntry.getValue(), map, next);
                 }
             }
-            treeRes.put(parentKey, next);
+            treeRes.merge(parentKey, next, (v1, v2) -> {
+                if (v1 instanceof Collection && v2 instanceof Collection) {
+                    ((Collection) v1).addAll((Collection) v2);
+                    return v1;
+                } else if (v1 instanceof Map && v2 instanceof Map) {
+                    ((Map) v1).putAll((Map) v2);
+                    return v1;
+                } else {
+                    return v2;
+                }
+            });
         } else {
             String key = String.valueOf(processorKey);
             treeRes.put(key, sourceAndMetadataMap.get(parentKey));
         }
     }
 
+    /**
+     * Process the nested key, such as "a.b.c" to "a", "b.c"
+     * @param nestedFieldMapEntry
+     * @return A pair of the original key and the target key
+     */
+    private Pair<String, Object> processNestedKey(final Map.Entry<String, Object> nestedFieldMapEntry) {
```

Comment:
Do we have a unit test for this?
I don't see any test for this method in `InferenceProcessorTests` file.

Also, this might break backward compatibility for user who is using dot in their key already. Are we fine on that?

Reply:
let me add the test. For the bwc we should be fine, it's more like an edge case if someone expecting "a.b.c" to be a single field and not the structure. It's one way or another and we're making this change based on requests from users and because this syntax supported in other places in Open search.

Reply:
Not about bwc test. Just writing it as unit test friendly would make it easier to maintain.

Reply:
Got it, added unit tests